### PR TITLE
Add io.github.wazzaps.Fingerpaint

### PR DESCRIPTION
**Pull request by Wazzaps**
### Please confirm your submission meets all the criteria

- [x] I have read the [App Requirements][reqs] and [App Maintenance][maint] pages.
- [x] My pull request follows the instructions at [App Submission][submission].
- [x] I am using only the minimal set of permissions. *(If not, please explain each non-standard permission.)*
- [x] All assets referenced in the manifest are redistributable by any party.  If not, the unredistributable parts are using an extra-data source type.
- [x] I am an upstream contributor to the project. If not, I contacted upstream developers about submitting their software to Flathub.
- [x] I own the domain used in the application ID or the domain has a policy for delegating subdomains (e.g. GitHub, SourceForge).
- [x] Any additional patches or files have been submitted to the upstream projects concerned.

[reqs]: https://docs.flathub.org/docs/for-app-authors/requirements
[maint]: https://docs.flathub.org/docs/for-app-authors/maintanance
[submission]: https://docs.flathub.org/docs/for-app-authors/submission

### CODE REVIEW

**Comment by Wazzaps** [comment]
bot, build com.github.wazzaps.Fingerpaint

**Comment by Wazzaps** [comment]
bot, build com.github.wazzaps.Fingerpaint

**Comment by Wazzaps** [comment]
bot, build io.github.wazzaps.Fingerpaint

**Comment by Wazzaps** [comment]
bot, build io.github.wazzaps.Fingerpaint

File: externals/PyGObject-3.44.1.tar.gz
Comment (hfiguiere) [inline externals/PyGObject-3.44.1.tar.gz:1]: you NEVER EVER add tarball to a flathub submission.

File: io.github.wazzaps.Fingerpaint.yml [hunk 1]
```diff
@@ -0,0 +1,45 @@
+app-id: io.github.wazzaps.Fingerpaint
+runtime: org.gnome.Platform
+runtime-version: '44'
+sdk: org.gnome.Sdk
+command: "/app/bin/fingerpaint"
```
Comment (hfiguiere) [inline io.github.wazzaps.Fingerpaint.yml:5]:
```suggestion
command: "fingerpaint"
```

File: io.github.wazzaps.Fingerpaint.yml [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+  - "--filesystem=/run/udev:ro"
+
+  # Access to dconf/gsettings for disabling touchpad under GNOME/Wayland
+  - "--filesystem=xdg-run/dconf:ro"
+  - "--filesystem=~/.config/dconf:ro"
+  - "--talk-name=ca.desrt.dconf"
+  - "--env=DCONF_USER_CONFIG_DIR=~/.config/dconf"
+modules:
+  - generated-poetry-sources.json
+  - name: fingerpaint
```
Comment (hfiguiere) [inline io.github.wazzaps.Fingerpaint.yml:25]: this should be the last module because its the main one.

File: io.github.wazzaps.Fingerpaint.metainfo.xml
Comment (hfiguiere) [inline io.github.wazzaps.Fingerpaint.metainfo.xml:1]: you can remove this file it's never referenced and it is in the repository upstream it seems

File: io.github.wazzaps.Fingerpaint.metainfo.xml [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    <url type="homepage">https://github.com/Wazzaps/fingerpaint</url>
+    <url type="help">https://github.com/Wazzaps/fingerpaint/issues</url>
+    <url type="bugtracker">https://github.com/Wazzaps/fingerpaint/issues</url>
+
+    <description>
+        <p>
+            This utility lets you draw using your laptop's touchpad.
+            Pressing any key or clicking the touchpad will finish the drawing.
+            Your touchpad will not control the cursor while drawing, though external mice should still work.
+            It has support for both X11 (all desktop environments), and Wayland (just Gnome at the moment).
```
Comment (hfiguiere) [inline io.github.wazzaps.Fingerpaint.metainfo.xml:20]: so what happen if the user runs KDE Plasma on top of wayland?

File: io.github.wazzaps.Fingerpaint.yml [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+  # Access to udev for finding raw touchpad device
+  - "--device=all"
+  - "--filesystem=/sys:ro"
+  - "--filesystem=/run/udev:ro"
+
+  # Access to dconf/gsettings for disabling touchpad under GNOME/Wayland
+  - "--filesystem=xdg-run/dconf:ro"
+  - "--filesystem=~/.config/dconf:ro"
+  - "--talk-name=ca.desrt.dconf"
+  - "--env=DCONF_USER_CONFIG_DIR=~/.config/dconf"
```
Comment (hfiguiere) [inline io.github.wazzaps.Fingerpaint.yml:22]: gsettings access is available by default.... you likely do NOT need this.

File: io.github.wazzaps.Fingerpaint.yml [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+        commit: "9d1eac5b87ce9c4c4034eae742bf12ae457cbf6d"
+
+  # Taken from: https://archlinux.org/packages/extra/x86_64/xorg-xinput/
+  - name: xinput
+    buildsystem: simple
+    build-commands:
+      - zstd -d < xorg-xinput-1.6.4-1-x86_64.pkg.tar.zst | tar x usr/bin/xinput && mv usr/bin/xinput /app/bin/
+    sources:
+      - type: file
+        path: "externals/xorg-xinput-1.6.4-1-x86_64.pkg.tar.zst"
```
Comment (hfiguiere) [inline io.github.wazzaps.Fingerpaint.yml:45]: so what happen when we build this on aarch64?

File: io.github.wazzaps.Fingerpaint.metainfo.xml (hunk 3 above)
Comment (Wazzaps) [reply io.github.wazzaps.Fingerpaint.metainfo.xml:20]: The tool fails to lock the cursor, though most functionality keeps working.

File: io.github.wazzaps.Fingerpaint.yml [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+  - "--env=DCONF_USER_CONFIG_DIR=~/.config/dconf"
+modules:
+  - generated-poetry-sources.json
+  - name: fingerpaint
+    buildsystem: simple
+    build-commands:
+      - python3 setup.py bdist_wheel && pip3 install --no-deps --no-index --prefix=/app dist/*.whl
+      - install -D flatpak/$FLATPAK_ID.metainfo.xml $FLATPAK_DEST/share/appdata/$FLATPAK_ID.appdata.xml
+      - install -D flatpak/$FLATPAK_ID.desktop $FLATPAK_DEST/share/applications/$FLATPAK_ID.desktop
+      - install -D snap/gui/fingerpaint.svg $FLATPAK_DEST/share/icons/$FLATPAK_ID.svg
```
Comment (hfiguiere) [inline io.github.wazzaps.Fingerpaint.yml:31]: this is not the correct path for install incons.
```suggestion
      - install -D snap/gui/fingerpaint.svg $FLATPAK_DEST/share/icons/hicolor/scalable/apps/$FLATPAK_ID.svg
```

File: externals/PyGObject-3.44.1.tar.gz
Comment (Wazzaps) [reply externals/PyGObject-3.44.1.tar.gz:1]: [This page](https://github.com/flathub/flathub/wiki/App-Requirements) says the build process cannot access the internet, so I vendored all dependencies in using the tool referenced there and downloaded them.

Do you suggest that keeping it as a URL will work?

File: io.github.wazzaps.Fingerpaint.yml (hunk 5 above)
Comment (Wazzaps) [reply io.github.wazzaps.Fingerpaint.yml]: Added a secondary source for aarch64

**Comment by Wazzaps** [comment]
bot, build io.github.wazzaps.Fingerpaint

**Comment by Wazzaps** [comment]
bot, build io.github.wazzaps.Fingerpaint

File: io.github.wazzaps.Fingerpaint.yml [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+  - "--device=dri"
+
+  # Access to udev for finding raw touchpad device
+  - "--device=all"
+  - "--filesystem=/sys:ro"
+  - "--filesystem=/run/udev:ro"
+modules:
+  - generated-poetry-sources.json
+
+  # Taken from: https://archlinux.org/packages/extra/x86_64/xorg-xinput/
```
Comment (hfiguiere) [inline io.github.wazzaps.Fingerpaint.yml:20]: you should build this from source.

File: generated-poetry-sources.json [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+            "sha256": "299db8628cc73b237fc1cc57d3c2948faa0756e2a58b6194b5bf81dc2081f1e3"
+        },
+        {
+            "type": "file",
+            "url": "https://files.pythonhosted.org/packages/e5/9f/541adc3a06ea28369c4ab52d2cf95597bb2ef626b3cf1c61afa26f735461/pycairo-1.22.0.tar.gz",
+            "sha256": "b34517abdf619d4c7f0274f012b398d9b03bab7adc3efd2912bf36be3f911f3f"
+        },
+        {
+            "type": "file",
+            "url": "https://files.pythonhosted.org/packages/44/80/96d6317a15a13a4f80ffa61118dc144a70756135fbc6ace30f9b033f51f7/PyGObject-3.44.1.tar.gz",
```
Comment (hfiguiere) [inline generated-poetry-sources.json:20]: this is in the GNOME runtime.

File: generated-poetry-sources.json [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+    ],
+    "sources": [
+        {
+            "type": "file",
+            "url": "https://files.pythonhosted.org/packages/05/50/629b011a7f61cb2fca754ea8631575784bf8605a1ec4d6970a010bc54e2b/evdev-1.6.1.tar.gz",
+            "sha256": "299db8628cc73b237fc1cc57d3c2948faa0756e2a58b6194b5bf81dc2081f1e3"
+        },
+        {
+            "type": "file",
+            "url": "https://files.pythonhosted.org/packages/e5/9f/541adc3a06ea28369c4ab52d2cf95597bb2ef626b3cf1c61afa26f735461/pycairo-1.22.0.tar.gz",
```
Comment (hfiguiere) [inline generated-poetry-sources.json:15]: this is in the GNOME runtime.

**Comment by hfiguiere** [comment]
```
Failed to download sources: module xinput: Wrong sha256 checksum for xorg-xinput-1.6.4-1-aarch64.pkg.tar.xz, expected "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855", was "bf630b44490486b3264bd80d7678a45e221ea185a513dc0a1ab5ef03e24c8bb5"
```

**Comment by Wazzaps** [comment]
bot, build io.github.wazzaps.Fingerpaint

**Comment by Wazzaps** [comment]
@hfiguiere Is there anything else I need to do?

File: io.github.wazzaps.Fingerpaint.yml [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+  - "--filesystem=/sys:ro"
+  - "--filesystem=/run/udev:ro"
+modules:
+  - generated-poetry-sources.json
+
+  - name: xinput
+    buildsystem: autotools
+    cleanup:
+      - '/share/man'
+      - '/lib/debug'
```
Comment (hfiguiere) [inline io.github.wazzaps.Fingerpaint.yml:24]:
```suggestion
```

**Comment by hfiguiere** [comment]
bot, build io.github.wazzaps.Fingerpaint

**Comment by hfiguiere** [comment]
This is terrible UX

image

File: io.github.wazzaps.Fingerpaint.yml [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+finish-args:
+  # GUI Permissions
+  - "--share=ipc"
+  - "--socket=wayland"
+  - "--socket=fallback-x11"
+  - "--device=dri"
+
+  # Access to udev for finding raw touchpad device
+  - "--device=all"
+  - "--filesystem=/sys:ro"
```
Comment (hfiguiere) [inline io.github.wazzaps.Fingerpaint.yml:15]: this doesn't seem to be necessary.

**Comment by hfiguiere** [comment]
The script content:
```shell
#!/usr/bin/env bash
set -e
echo 'ENV{ID_INPUT_TOUCHPAD}=="1", MODE="0664"' > /etc/udev/rules.d/99-touchpad-access.rules
udevadm control --reload-rules
udevadm trigger
if which snap > /dev/null; then
  snap connect fingerpaint:hardware-observe 2>/dev/null && \
  snap connect fingerpaint:raw-input 2>/dev/null || true
fi
echo 'Done! Run `fingerpaint` again.' 1>&2
```

so basically THIS will modify the system. This is really not nice.

File: io.github.wazzaps.Fingerpaint.yml (hunk 11 above)
Comment (Wazzaps) [reply io.github.wazzaps.Fingerpaint.yml:15]: I need to enumerate the touchpad(s), only reasonable way to do this is with udev, which requires /sys access.

**Comment by Wazzaps** [comment]
^ If you have a better idea, give it to me.
Also, I do this as a sideproject, I only have a couple of hours per week to work on this. Don't expect corporate-level polish.
And I'd rather spend my limited time adding features, not worrying about packaging and distribution.

**Comment by hfiguiere** [comment]
> Don't expect corporate-level polish.

I think you are misunderstanding who the contributors to flathub are.

File: io.github.wazzaps.Fingerpaint.yml (hunk 11 above)
Comment (hfiguiere) [reply io.github.wazzaps.Fingerpaint.yml:15]: it did find it without /sys.

also it should be noted that the use of udev that way is not guaranteed at all.

**Comment by Wazzaps** [comment]
Other app stores (Google Play, Canonical Snapcraft, etc.) allow and encourage independent developers, do you expect only corporations to submit apps? For a platform which has a rounding-error market share? And a culture of not paying for apps?
I work on open source projects as a passion, and this interaction is kinda killing that passion.

When you say "you NEVER EVER add tarball to a flathub submission.", that tone feels like an attack. Instead of writing that, maybe clear up the documentation? Or try to understand why I would do such a thing? Instead I feel like my boss is shouting at me for losing $50 million dollars. Try to see the human on the other side.

**Comment by hfiguiere** [comment]
outtage should last 12 more hours, fyi.

**Comment by hfiguiere** [comment]
bot, build io.github.wazzaps.Fingerpaint

**Comment by Wazzaps** [comment]
> The `which snap` part should be changed not to output an error when it's not found, or even better, the script changed to take a parameter and avoid running snap commands at all.

I think I'll duplicate the script for snap, and remove the snap part entirely from the Flatpak version.

**Comment by Wazzaps** [comment]
bot, build io.github.wazzaps.Fingerpaint

**Comment by Wazzaps** [comment]
@barthalion Anything else blocking release?

File: io.github.wazzaps.Fingerpaint.yml [hunk 12]
```diff
@@ -0,0 +1,41 @@
+app-id: io.github.wazzaps.Fingerpaint
+runtime: org.gnome.Platform
+runtime-version: '44'
```
Comment (bbhtt) [inline io.github.wazzaps.Fingerpaint.yml:3]:
```suggestion
runtime-version: '46'
```

44 is EOL

**Comment by bbhtt** [comment]
bot, build io.github.wazzaps.Fingerpaint

**Comment by bbhtt** [comment]
Please add a developer name https://docs.flathub.org/docs/for-app-authors/metainfo-guidelines/#name-summary-and-developer-name and a lauchable tag https://docs.flathub.org/docs/for-app-authors/metainfo-guidelines/#launchable to appdata file.

**Comment by Wazzaps** [comment]
bot, build io.github.wazzaps.Fingerpaint

**Comment by Wazzaps** [comment]
bot, build io.github.wazzaps.Fingerpaint

**Comment by Wazzaps** [comment]
bot, build io.github.wazzaps.Fingerpaint

**Comment by bbhtt** [comment]
You have to update the appdata file then update the [commit here](https://github.com/flathub/flathub/pull/4429/files#diff-0c54e51c0d844a117189fd4e0ddd335cb129e741d7fa49c809080f360cebcc98R41), otherwise it won't work.

**Comment by Wazzaps** [comment]
bot, build io.github.wazzaps.Fingerpaint
(sorry, had git problems)

**Comment by bbhtt** [comment]
Is it working for you? I don't have the hardware to test.

**Comment by Wazzaps** [comment]
It worked last time I tested it, unfortunately I’m not near my laptop to retest.

**Comment by Wazzaps** [comment]
@bbhtt Can confirm that it works (installed using the command above):
Screenshot from 2024-05-11 18-34-11

**Comment by Wazzaps** [comment]
@bbhtt Anything left blocking release? :slightly_smiling_face:

**Comment by bbhtt** [comment]
We batch the merges and it's a weekend, you'll have to wait at least until tomorrow.

**Pull request closed without merging**